### PR TITLE
Deepmemory support

### PR DESCRIPTION
**Pull request by adolkhan**
# Description

In this PR, we introduce DeepLake's new feature, DeepMemory. This feature trains a compact network to align the embedding spaces of user queries with corpus data. With DeepMemory, we achieve a 15% improvement in hit rate, as detailed in the attached Jupyter notebook.

Fixes # (issue)

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [ ] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: docs/optimizing/advanced_retrieval/How to get +15% hit_rate improvement using DeepMemory.ipynb [hunk 1]
```diff
@@ -0,0 +1,749 @@
+{
```
Comment (hatianzhang) [inline docs/optimizing/advanced_retrieval/How to get +15% hit_rate improvement using DeepMemory.ipynb:1]: <p>some nitpicks and mispelling</p><p>like <code>relevence</code> to relevance</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/run-llama/llama_index/pull/8625/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (hatianzhang) [inline docs/optimizing/advanced_retrieval/How to get +15% hit_rate improvement using DeepMemory.ipynb:1]: <p>can you mind showing head of train/text examples in the notebook to tell users what is the format like</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/run-llama/llama_index/pull/8625/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (hatianzhang) [inline docs/optimizing/advanced_retrieval/How to get +15% hit_rate improvement using DeepMemory.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #7.            relevance.append([(relevant_docs[doc][0], 1)])
</pre><p>so relevance here is fixed as 1.</p><p>is it possible to send graded relevance</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/run-llama/llama_index/pull/8625/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (hatianzhang) [inline docs/optimizing/advanced_retrieval/How to get +15% hit_rate improvement using DeepMemory.ipynb:1]: <p>the results look great.</p><p>probably can add more metrics like precision, MAP, NDCG in future with different top_k, where k at 5, 15, 20 so on</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/run-llama/llama_index/pull/8625/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

**Comment by hatianzhang** [comment]
one thing high level is that you split the train and test set on the same document.
One document usually have similar word distribution. The model could be a little biased towards this single document.
Would be interested to see how deep memory performs on multiple documents.

**Comment by adolkhan** [comment]
> one thing high level is that you split the train and test set on the same document.
> One document usually have similar word distribution. The model could be a little biased towards this single document.
> Would be interested to see how deep memory performs on multiple documents.

Nope, there are multiple documents, I am parsing whole deeplake docs (it is 120 different urls and constructing synthetic queries on this docs) So I would say we are doing search and training across multiple documents. Will address all other comments in a bit

File: docs/optimizing/advanced_retrieval/How to get +15% hit_rate improvement using DeepMemory.ipynb (hunk 1 above)
Comment (adolkhan) [reply docs/optimizing/advanced_retrieval/How to get +15% hit_rate improvement using DeepMemory.ipynb:1]: not yet

Comment (adolkhan) [reply docs/optimizing/advanced_retrieval/How to get +15% hit_rate improvement using DeepMemory.ipynb:1]: added printings

Comment (adolkhan) [reply docs/optimizing/advanced_retrieval/How to get +15% hit_rate improvement using DeepMemory.ipynb:1]: thanks a lot for suggestions, we will definitely add them later

**Review by hatianzhang — APPROVED** [review]

**Pull request merged**